source/server/server.cc

```diff
@@ -614,23 +614,21 @@ void InstanceImpl::notifyCallbacksForStage(Stage stage, Event::PostCb completion
     }
   }
 
+  // Wrap completion_cb so that it only gets invoked when all callbacks for this stage
+  // have finished their work.
+  std::shared_ptr<Event::PostCb> cb_guard(new Event::PostCb([] {}),
+                                          [this, completion_cb](Event::PostCb* cb) {
+                                            ASSERT(std::this_thread::get_id() == main_thread_id_);
```

Comment:
I hadn't seen the original PR, but why are we not using the thread-factory abstraction here? We have easy access to it via server_->api..

Reply:
No special reason - I think I just copied a similar pattern used in [TLS](https://github.com/envoyproxy/envoy/blob/0e12efc140430033b98c9927464df2258c2aeffd/source/common/thread_local/thread_local_impl.cc#L83)